File: cpp/src/edu/uf/interactable/Macrophage.cpp
/*
 * Macrophage.cpp
 *
 *  Created on: Jul 17, 2024
 *      Author: henriquedeassis
 */

#include "Macrophage.h"
#include "../utils/Id.h"
#include "../utils/Constants.h"
#include "../utils/Rand.h"

namespace edu {
namespace uf {
namespace interactable {

using namespace edu::uf::utils;

const int Macrophage::classId = Id::getId();
const std::string Macrophage::NAME = "Macrophage";
std::string Macrophage::chemokine = "";
int Macrophage::totalCells = 0;
double Macrophage::totalIron = 0;
int Macrophage::interactionId = utils::Id::getId();

Macrophage::Macrophage(double ironPool, intracellularState::IntracellularModel* network)
    : Leukocyte(ironPool, network), maxMoveStep(-1), engaged(false) {
    Macrophage::totalCells++;
    Macrophage::totalIron += ironPool;
}

int Macrophage::getClassId() const {
	return classId;
}

void Macrophage::setMaxMoveStep(int moveStep) {
    maxMoveStep = moveStep;
}

int Macrophage::getInteractionId() const {
    return interactionId;
}

bool Macrophage::isTime() const {
    return getClock()->toc();
}

std::string Macrophage::getChemokine() {
    return chemokine;
}

void Macrophage::setChemokine(const std::string& chemokine) {
    Macrophage::chemokine = chemokine;
}

int Macrophage::getTotalCells() {
    return totalCells;
}

void Macrophage::setTotalCells(int totalCells) {
    Macrophage::totalCells = totalCells;
}

double Macrophage::getTotalIron() {
    return totalIron;
}

void Macrophage::setTotalIron(double totalIron) {
    Macrophage::totalIron = totalIron;
}

bool Macrophage::isEngaged() const {
    return engaged;
}

void Macrophage::setEngaged(bool engaged) {
    this->engaged = engaged;
}

int Macrophage::getMaxMoveSteps() {
    double r = 1.0;
    if (maxMoveStep == -1) {
        maxMoveStep = utils::Rand::getRand().randpois(utils::constexprants::MA_MOVE_RATE_ACT * r);
    }
    return maxMoveStep;
}

bool Macrophage::templateInteract(Interactable* interactable, int x, int y, int z) {
    return interactable->interact(this, x, y, z);
}

void Macrophage::incIronPool(double qtty) {
    setIronPool(getIronPool() + qtty);
    Macrophage::totalIron += qtty;
}

void Macrophage::die() {
    if (getBooleanNetwork()->getState(intracellularState::IntracellularModel::LIFE_STATUS) != intracellularState::IntracellularModel::DEAD) {
        getBooleanNetwork()->setState(intracellularState::IntracellularModel::LIFE_STATUS, intracellularState::IntracellularModel::DEAD);
        Macrophage::totalCells--;
    }
}

std::string Macrophage::attractedBy() const {
    return chemokine;
}

int Macrophage::getMaxCell() {
    return utils::constexprants::MA_MAX_CONIDIA;
}

std::string Macrophage::getName() const {
    return NAME;
}

} /* namespace interactable */
} /* namespace uf */
} /* namespace edu */

File: java/src/edu/uf/interactable/Macrophage.java
package edu.uf.interactable;

import edu.uf.intracellularState.IntracellularModel;
import edu.uf.utils.Constants;
import edu.uf.utils.Id;
import edu.uf.utils.Rand;


public class Macrophage extends Leukocyte{

    public static final String NAME = "Macrophage";

    private static String chemokine = null;
    
    private static int totalCells = 0;
    private static double totalIron = 0; 
    
    private int maxMoveStep;
    private boolean engaged;
    private static int interactionId = Id.getId();
    
    
    
	public Macrophage(double ironPool, IntracellularModel network) {
    	super(ironPool, network);
    	Macrophage.totalCells = Macrophage.totalCells + 1;
        this.maxMoveStep = -1; 
        Macrophage.totalIron = Macrophage.totalIron + ironPool;
    }
	
	public void setMaxMoveStep(int moveStep) {
		this.maxMoveStep = moveStep;
	}
	
	public int getInteractionId() {
		return interactionId;
	}
	
	public boolean isTime() {
		return this.getClock().toc();
	}
    
    public static String getChemokine() {
		return chemokine;
	}

	public static void setChemokine(String chemokine) {
		if(chemokine == null)
			Macrophage.chemokine = chemokine;
	}

	public static int getTotalCells() {
		return totalCells;
	}

	public static void setTotalCells(int totalCells) {
		Macrophage.totalCells = totalCells;
	}

	public static double getTotalIron() {
		return totalIron;
	}

	public static void setTotalIron(double totalIron) {
		Macrophage.totalIron = totalIron;
	}

	public boolean isEngaged() {
		return engaged;
	}

	public void setEngaged(boolean engaged) {
		this.engaged = engaged;
	}

    public int getMaxMoveSteps() { 
    	double r = 1.0;
    	//if(this.getExternalState() == 1)r = Constants.NET_COUNTER_INHIBITION;
        if(this.maxMoveStep == -1) {
        	//this.maxMoveStep = Rand.getRand().randunif() < Constants.MA_MOVE_RATE_ACT ? 1 : 0;
        	this.maxMoveStep = Rand.getRand().randpois(Constants.MA_MOVE_RATE_ACT*r);
        }
            //
        return this.maxMoveStep;
    }

    protected boolean templateInteract(Interactable interactable, int x, int y, int z) {
        return interactable.interact(this, x, y, z);
    }
        
    public void incIronPool(double qtty) {
        this.setIronPool(this.getIronPool() + qtty);
        Macrophage.totalIron = Macrophage.totalIron + qtty;
    }

    public void die() {
    	if(this.getBooleanNetwork().getState(IntracellularModel.LIFE_STATUS) != IntracellularModel.DEAD) {
    		this.getBooleanNetwork().setState(IntracellularModel.LIFE_STATUS, IntracellularModel.DEAD);
            Macrophage.totalCells = Macrophage.totalCells - 1;
        }
    }

    public String attractedBy() {
        return Macrophage.chemokine;
    }

	@Override
	public int getMaxCell() {
		return Constants.MA_MAX_CONIDIA;
	}
	
	public String getName() {
    	return NAME;
    }
	
}
